src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -213,10 +214,14 @@ void reconnect() throws TimeoutException, ExecutionException, InterruptedExcepti
         // For subsequent connects, the client connects with cleanSession=false
         CompletableFuture<Void> voidCompletableFuture =  CompletableFuture.completedFuture(null);
         if (initialConnect.get()) {
-            voidCompletableFuture = establishConnection(true).thenCompose((session) -> {
-                initialConnect.set(false);
-                return disconnect();
-            });
+            try {
+                voidCompletableFuture = establishConnection(true).thenCompose((session) -> {
+                    initialConnect.set(false);
+                    return disconnect();
+                });
+            } catch (RuntimeException e) {
+                voidCompletableFuture.completeExceptionally(e);
+            }
```

Comment:
instead, add the catch inside establishConnection and return a completed future which is completed exceptionally.

establishConnection is called twice, so your try catch won't catch everything.

Reply:
that makes sense, thanks. Fixed it.